docs/Blog-Post.md

```diff
@@ -0,0 +1,578 @@
+is performed on different machines.
+
+Levanter works out of the box with either [Hugging Face Datasets](https://huggingface.co/datasets) (including streaming) or urls of (compressed)
+jsonl files. Caches can be stored in any fsspec-compatible file system, including GCS and local file systems. We use
+[Hugging Face Tokenizers](https://huggingface.co/docs/tokenizers/) for tokenization.
+
+### Live Visualization during Training
+
+Levanter also provides a feature for visualizing the probability of each token in a sample of the validation set during training.
+When training large models, it can be difficult to get a sense of how the model is learning. This is especially true
+when training on novel data sets. As an example, we have seen issues with early versions of new datasets where the
+model had suspiciously low loss.
+
+The live visualization produces a heatmap of the log probability of each token in a sample of the validation set
+that is updated periodically during training. Here is an example of the token probability visualization in action on a
+small, quick training run:
+
+![video showing heat map of token probabilities for a sample of the validation set evolving as training progresses](figures/token_probabilities.mov)
+
+The darker, more purple the color, the lower the probability of the token. The lighter, more yellow the color, the higher the probability.
+This visualization is logged to WandB as training progresses and can be viewed interactively. We have found this to be a
+nice alternative to just staring obsessively at the loss curve (not that we ever do that).
+
+In the past, we have used our visualization to identify a pattern of highly but not perfectly redundant data in a new dataset
+(what we call "madlib duplicates"), suggesting that the model is "wasting" time and context length on low-value data.
+We've also used it to qualitatively assess how alternative architectures (like [Backpacks](http://backpackmodels.science/))
+learn differently from Transformers.
+
+### Other Features
+
+* **Training**: Levanter uses [Optax](https://github.com/deepmind/optax) for optimization
+  (though our new optimizer, [Sofia](https://arxiv.org/abs/2305.14342), is coming to Levanter soon!)
+* **Logging**: Logging is done with [WandB](https://wandb.ai/), complete with a fancy online visualization of the validation set during training.
+* **Checkpointing**: Distributed checkpointing is supported via Google's [TensorStore](https://google.github.io/tensorstore/) library. Training can even be resumed on a different number of hosts, though this breaks reproducibility for now.
+* **Export**: We also support exporting models to the Hugging Face Hub, with export compatible with Pytorch and Transformers via [SafeTensors](https://github.com/huggingface/safetensors).
+* **Stability**: The GPT-2 implementation uses the [Mistral stability trick](https://crfm.stanford.edu/2021/08/26/mistral.html) to improve stability during training.
```

Comment:
Will this be no longer necessary with Sophia?

Reply:
I don't see why, but I'm of course not optimization person. My understanding is that the issue is that the grad norm gets big and then numerics get wonky. Is your point that the hessian estimate will say "whoa this is steep, slow down"?

I think update clipping is helpful here, as is more layer norms. (always want more layer norms). I guess sophia and update clipping are doing similar things in this regard.

 it's also usually not a huge problem with bf16 (though we saw a few problems with backpacks even with the tweak...)